### PR TITLE
Derivations for (Co)Equalizers

### PR DESCRIPTION
**Pull request by TKuh**
And additionally make the derivations of FiberProduct / Pushout rely on it. See https://github.com/homalg-project/CAP_project/issues/1076

Remarks:
- Right now Abelian categories are not constructively Abelian, because the current [derivation](https://github.com/homalg-project/CAP_project/blob/ecddb9abe70e0e8c465a1249d723454362222805/CAP/gap/DerivedMethods.gi#L2721) of `(Co)Equalizer` is blocked by the filter for limits and I first want to ask about the proper way to change it.
- I renamed `DirectSumDiagonalDifference` to `Shifted(Co)DiagonalDifferenceOfMorphisms` since the original name wasn't really fitting, and I called the new version for equalizers `(Co)TelescopingDifferenceOfMorphisms` (in analogy to telescoping sums). Proposals for better names? :)
-  The [compilation](https://github.com/TKuh/CAP_project/tree/FiberProduct_new_derivation) of the new derivation path `FiberProduct` -> `ShiftedDiagonalDifferenceOfMorphisms` -> `TelescopingDifferenceOfMorphisms` gives an equivalent result to the former `FiberProduct` -> `DirectSumDiagonalDifference`. However, the compiled old derivation was shorter.

### CODE REVIEW

**Comment by zickgraf** [comment]
Very nice, thanks!

Regarding your remarks:

1. Thanks for noticing this, fixed: https://github.com/homalg-project/CAP_project/pull/1126.
2. I have to think about this.
3. We can easily fix this with suitable logic templates, but we should first understand where the fixed list accesses (`list[1]` and `list[2]`) come from and see if those can be avoided.

Regarding your first commit: One may not derive `EmbeddingOfEqualizer` as `KernelEmbedding( TelescopingDifferenceOfMorphisms )` directly because the user might already have installed `Equalizer` in a different, possibly incompatible, way. The solution for this problem is the following: Introduce `IsomorphismFromEqualizerToKernelOfTelescopingDifferenceOfMorphisms` which translates between a user-given implementation of the equalizer and the derivation as the kernel of the telescoping difference, and set this to the identity in a final derivation only in case the user has not installed any of the operations concerning equalizers (cf. `IsomorphismFromFiberProductToKernelOfDiagonalDifference`).

I think we best go through this commit by commit.

**Comment by TKuh** [comment]
>     3. We can easily fix this with suitable logic templates, but we should first understand where the fixed list accesses (`list[1]` and `list[2]`) come from and see if those can be avoided.

This is probably from `Source( diagram[ 1 ] )` and `Source( diagram[ 2 ] )`.
https://github.com/homalg-project/CAP_project/blob/15c31281800b77dcc7bf2d71ba940b03af0ab678/CAP/gap/DerivedMethods.gi#L1929-L1931
We could give `TelescopingDifferenceOfMorphisms` the common source of all the morphisms in `diagram` directly, then the list access might be prevented? `Equalizer`, `EmbeddingOfEqualizer`, etc. do it, too.

> Regarding your first commit: One may not derive `EmbeddingOfEqualizer` as `KernelEmbedding( TelescopingDifferenceOfMorphisms )` directly because the user might already have installed `Equalizer` in a different, possibly incompatible, way. The solution for this problem is the following: Introduce `IsomorphismFromEqualizerToKernelOfTelescopingDifferenceOfMorphisms` which translates between a user-given implementation of the equalizer and the derivation as the kernel of the telescoping difference, and set this to the identity in a final derivation only in case the user has not installed any of the operations concerning equalizers (cf. `IsomorphismFromFiberProductToKernelOfDiagonalDifference`).

Thank you, now I understand these isomorphisms. :) 
I gave it a try.

**Comment by zickgraf** [comment]
> We could give TelescopingDifferenceOfMorphisms the common source of all the morphisms in diagram directly, then the list access might be prevented? Equalizer, EmbeddingOfEqualizer, etc. do it, too.

Yes, that's the way to go to also support the case of an empty equalizer diagram. See here https://en.wikipedia.org/wiki/Equaliser_(mathematics)#In_category_theory for the background why the source is needed.

**Comment by zickgraf** [comment]
To record what we have discussed on Friday:

1. The name `DirectSumDiagonalDifference` should be kept.
2. `TelescopeDifferenceOfMorphisms` should be renamed to `JointPairwiseDifferencesOfMorphisms`.
3. All derivations should work without relying on concrete list access (i.e. `list[1]`).

**Comment by mohamed-barakat** [comment]
Could you please rebase this branch?

**Comment by TKuh** [comment]
I rebased, but I have not yet worked on the todo's above (I plan to do it either later today or tomorrow).

**Comment by TKuh** [comment]
There are compiler warnings with `LinearAlgebraForCAP`, which I don't know how to remove, but otherwise this is ready for review.

As discussed, there are methods which can be switched from `DirectSum` to `DirectProduct/Coproduct`, but I would like to do this in a separate PR (unless you prefer it here).

File: CAP/examples/TerminalCategoryWithMultipleObjects.g [hunk 1]
```diff
@@ -8,8 +8,7 @@ LoadPackage( "MonoidalCategories" );
 T := TerminalCategoryWithMultipleObjects( );
 #! TerminalCategoryWithMultipleObjects( )
 InfoOfInstalledOperationsOfCategory( T );
-#! 63 primitive operations were used to derive 280 operations for this category
-#! which algorithmically
+#! 63 primitive operations were used to derive 304 operations for this category which algorithmically
```
Comment (mohamed-barakat) [inline CAP/examples/TerminalCategoryWithMultipleObjects.g:11]:
```suggestion
#! 63 primitive operations were used to derive 304 operations for this category
#! which algorithmically
```

**Comment by mohamed-barakat** [comment]
I rebased on PR https://github.com/homalg-project/CAP_project/pull/1197 and resolved the conflicts. I can push onto your branch once the mentioned PR is merged.

**Comment by mohamed-barakat** [comment]
> There are compiler warnings with `LinearAlgebraForCAP`, which I don't know how to remove, but otherwise this is ready for review.

I was unable to quickly locate the cause of the warning.

**Comment by zickgraf** [comment]
Thanks for the progress! I think all that remains is to remove `DirectSumDiagonalDifference` and `DirectSumCodiagonalDifference`. Those are still used in the derivations of FreydCategoriesForCAP, but with a list of fixed length 2.

I suggest to simply replace the calls to `DirectSumDiagonalDifference` in FreydCategoriesForCAP by the code you have written here: https://github.com/homalg-project/CAP_project/pull/1124/commits/3ce90e2a9cd1dad23b1b5cc2fdc54c982a975968#diff-ba2f61a1820f9344a717236211fac2426be05ccd6340fc34e6d4dfa1b81fbc8aR2298-R2309

**Comment by TKuh** [comment]
> Thanks for the progress! I think all that remains is to remove `DirectSumDiagonalDifference` and `DirectSumCodiagonalDifference`. 

The operations `IsomorphismFromKernelOfDiagonalDifferenceToFiberProduct` and `IsomorphismFromFiberProductToKernelOfDiagonalDifference` should also be removed, right?

> I suggest to simply replace the calls to `DirectSumDiagonalDifference` in FreydCategoriesForCAP by the code you have written here: [3ce90e2#diff-ba2f61a1820f9344a717236211fac2426be05ccd6340fc34e6d4dfa1b81fbc8aR2298-R2309](https://github.com/homalg-project/CAP_project/commit/3ce90e2a9cd1dad23b1b5cc2fdc54c982a975968#diff-ba2f61a1820f9344a717236211fac2426be05ccd6340fc34e6d4dfa1b81fbc8aR2298-R2309)

Thank you for this suggestion, this prevented a second question I would have had otherwise. :)

**Comment by zickgraf** [comment]
> The operations `IsomorphismFromKernelOfDiagonalDifferenceToFiberProduct` and `IsomorphismFromFiberProductToKernelOfDiagonalDifference` should also be removed, right?

Correct :-)

**Comment by zickgraf** [comment]
(and all the duals too, of course)

**Comment by zickgraf** [comment]
I think with https://github.com/homalg-project/CAP_project/pull/1237, all the parts are merged, thanks for the work!

**Pull request closed without merging**